lib/retry.ts

```diff
@@ -122,6 +137,11 @@ export async function retry<T>(config: RetryConfig<T>): Promise<T> {
       if (!err.translated) {
         err = translate(err);
       }
+      const isConnected = await checkNetworkConnection();
+      if (!isConnected) {
+        err.name = "ConnectionDetachError";
+        err.retryable = true;
```

Comment:
We make this check for every error seen by every `retry` attempt. Can we reduce the number of checks being made?
One example is when the existing error is already retryable, then we don't need to make this check.

Also, maybe we should only add this check when we see the `ServiceCommunicationError` which maps to the ENOTFOUND. Thoughts?

Reply:
`One example is when the existing error is already retryable, then we don't need to make this check` - This makes sense. I'll add one more check that if existing error is already retryble then no need to do anything.

`Also, maybe we should only add this check when we see the ServiceCommunicationError which maps to the ENOTFOUND. `  - This check would be really specific for our use case only. In future if user will get different non-retryable error then he will not be able to retry again.

Also, I believe that if the error is not retryable and if the network connection is down, we should anyways retry the operation. What do you think ?

Reply:
If the network was down, then we wouldn't be getting any error from the service in the first place isn't it?

Reply:
So you mean that we **always** get the same `ServiceCommunicationError`  during a network failure ? Doesn't it depends on the program state when network goes down ?

Also, what I am proposing above is that it doesn't matter whether we always get the ServiceCommunicationError or not, we should retry if the error is not retryable and we are not connected to the internet.

Reply:
If the network is down and we make a network request, Nodejs will always give ENOTFOUND error and amqp-common translates that to `ServiceCommunicationError`. Nodejs doesn't care about the program state, it only sees a network request being made.

ENOTFOUND can happen due to multiple reasons, network being down is one of them.

Reply:
I agree with above, but I've a couple of questions:
* Is there any specific reason to check for `ServiceCommunicationError` in this retry logic ? If we have figured out that the network is down and the error is not retryable, then we should just retry without having to worry about what kind of error was thrown. 
* I agree that the Nodejs will always give ENOTFOUND when making network request. But what about the cases when the connection is already established but failure happened while waiting for response? The other concern I have is that our check heavily relies on the fact that we always get ENOTFOUND on network failures. What if this implementation changes in future?

Reply:
> Is there any specific reason to check for ServiceCommunicationError in this retry logic ? If we have figured out that the network is down and the error is not retryable, then we should just retry without having to worry about what kind of error was thrown.

Say the retry logic failed because of some error in the sdk itself. In that case, the error will not be retryable (Only the amqp errors are mapped as retryable errors in amqp-common). In this case, we will be unnecessarily making the `dns.resolve` call.
Say that in this case, the network is also down, now we will be retrying even though the operation will never succeed because of error in our code.

> But what about the cases when the connection is already established but failure happened while waiting for response

Can you elaborate the case here? Failure at which point? In the service?

> The other concern I have is that our check heavily relies on the fact that we always get ENOTFOUND on network failures. What if this implementation changes in future

I would like to believe that the error code is standard enough that implementations won't change, but given that the [Nodejs Error docs](https://nodejs.org/api/errors.html) don't really list this error code, I am not a 100% sure.

It is worth our time looking at different kinds of network failures and how Nodejs reacts to each.

@bterlson Thoughts on Nodejs changing implementation on a non standard error code like ENOTFOUND?

Reply:
>Say the retry logic failed because of some error in the sdk itself. In that case, the error will not be retryable (Only the amqp errors are mapped as retryable errors in amqp-common). In this case, we will be unnecessarily making the dns.resolve call.
Say that in this case, the network is also down, now we will be retrying even though the operation will never succeed because of error in our code.

I agree with the above that this could happen but would think as an edge case in our scenario since clients would mostly require network communication for their tasks.

> But what about the cases when the connection is already established but failure happened while waiting for response.
Can you elaborate the case here? Failure at which point? In the service?

I was talking about the case when we have made connection with the service, sent a request and is now waiting for a response but network fails. What would happen in this case? Will we get the same ENOTFOUND error?

Overall, I think that adding `ServiceCommunicationError` may led our check to become too restrict or dependent upon implementation, but if we feel that we should add this check, I can make this change.

Reply:
> I agree with the above that this could happen but would think as an edge case in our scenario since clients would mostly require network communication for their tasks.

It is an edge case as in, it will happen only because the code we have written fails and we have a lot of lines of code that doesnt deal with network calls and each of them is a point of failure :) We deal with data transformation (in the servicebusmessage.ts file), events, event handlers, timers and so much more

> I was talking about the case when we have made connection with the service, sent a request and is now waiting for a response but network fails. What would happen in this case? Will we get the same ENOTFOUND error?

In this case, rhea will send a `disconnected` event first. This results in calling `detached` for all open senders and receivers. The `detached` function sees that the sdk didnt close the sender/receiver and will try to reconnect the sender/receiver. The first `retry` attempt to run `_init` to reconnect them will fail with the ServiceCommunicationError and that is when your network connection check will come in.

> Overall, I think that adding ServiceCommunicationError may led our check to become too restrict or dependent upon implementation

Yes, restricting this indeed my intention :). We need this check only to circumvent the problem of not being able to differentiate network connection being down and all other reasons that might result in `ServiceCommunicationError`. 

Regarding the dependency on the implementation and Nodejs deciding to send some other error than ENOTFOUND for network being down, I'll defer to @bterlson's expertise

Reply:
Ok, then I'll add one more check for `ServiceCommunicationError` with the `if (!isConnected)`. :)

Reply:
This is also something other SDKs will face.
@bterlson Can we have a note on this case of Nodejs not able to differentiate host not found vs network being down in the guidelines?

Reply:
Just one more concern about this :)
>I was talking about the case when we have made connection with the service, sent a request and is now waiting for a response but network fails. What would happen in this case? Will we get the same ENOTFOUND error?

>In this case, rhea will send a disconnected event first. This results in calling detached for all open senders and receivers. The detached function sees that the sdk didnt close the sender/receiver and will try to reconnect the sender/receiver. The first retry attempt to run _init to reconnect them will fail with the ServiceCommunicationError and that is when your network connection check will come in.


This case is specific for our Event Hub and Service bus SDK. If other user will use the same retry method of Amqp-common, will they get the same ENOTFOUND error?
In this case I was thinking It should either be `ConnectionTimeout` or `SocketClose` error?

Reply:
Re: ENOTFOUND, neither Node docs, errno docs, or getaddrinfo docs say anything about this error. However, a search of GitHub code shows we are not the only ones responding to this kind of error, so I find it extremely unlikely Node would change this error in the future. I've sent some tweets on the topic, will get back here if I hear anything different.

Reply:
ENOTFOUND is indeed not a real posix error and that explains why it's hard to find any info on it. It's actually a node-specific error that isn't documented. The source of the error is [here](https://github.com/nodejs/node/blob/e3d4a7d9994ff7d1bd144779f8b91d8302112bc0/lib/internal/errors.js#L435). While the comment implies the error will go away, [on twitter](https://twitter.com/addaleax/status/1103401216736980998), Anna (Node core member) suggests the error won't change and agrees that the comment should be removed and the error documented. I can push on that if we find it important, but personally I'm satisfied that this behavior won't change because of all the things that would break.

Reply:
> In this case I was thinking It should either be ConnectionTimeout or SocketClose error?

Are you saying that we should rename `ServiceCommunicationError` or the `ConnectionDetachedError`?

Reply:
Closing the loop here based on offline conversation

> This case is specific for our Event Hub and Service bus SDK. If other user will use the same retry method of Amqp-common, will they get the same ENOTFOUND error?

amqp-common is tightly coupled with rhea-promise and rhea. So, all users of amqp-common **will** get the `disconnected` event when the network goes down, which they can then react to just how we do.

Also, lets rename `ConnectionDetachedError` to `ConnectionLostError`